src/arsenic/services.py

```diff
@@ -1,45 +1,71 @@
 import abc
+import asyncio
 import os
+from asyncio.subprocess import DEVNULL
+from functools import partial
 from typing import List, TextIO
 
 import attr
+from aiohttp import ClientSession
 
 from arsenic.connection import Connection, RemoteConnection
-from arsenic.engines import Request, Engine, Auth, BasicAuth
 from arsenic.utils import free_port
 from arsenic.webdriver import WebDriver
+from arsenic.http import Auth, BasicAuth
 
 
-async def subprocess_based_service(engine: Engine,
-                                   cmd: List[str],
+async def stop_process(process):
+    process.terminate()
+    try:
+        await asyncio.wait_for(process.communicate(), 1)
+    except asyncio.futures.TimeoutError:
+        process.kill()
+    try:
+        await asyncio.wait_for(process.communicate(), 1)
+    except asyncio.futures.TimeoutError:
+        pass
+
+
+def sync_factory(func):
+    async def sync():
+        func()
+    return sync
+
+
+async def tasked(coro):
+    return await asyncio.get_event_loop().create_task(coro)
+
+
+async def subprocess_based_service(cmd: List[str],
                                    service_url: str,
                                    log_file: TextIO) -> WebDriver:
     closers = []
     try:
-        process = await engine.start_process(
-            cmd,
-            os.environ,
-            log_file
-        )
-        closers.append(process.close)
-        session = await engine.http_session()
-        closers.append(session.close)
-        request = Request(
-            url=service_url + '/status',
-            method='GET'
+        if log_file is os.devnull:
+            log_file = DEVNULL
+        process = await asyncio.create_subprocess_exec(
+            *cmd,
+            stdout=log_file,
+            stderr=log_file,
         )
+        closers.append(partial(stop_process, process))
+        session = ClientSession()
+        closers.append(sync_factory(session.close))
         count = 0
         while True:
             try:
-                await session.request(request)
+                await tasked(session.request(
+                    url=service_url + '/status',
+                    method='GET'
+                ))
                 break
             except:
+                # TODO: make this better
```

Comment:
I'm not sure how this would work. Basically we need to check if the webdriver is running (by checking it's `/status` endpoint). The first few requests might fail because the webdriver HTTP server is not running yet, which is why I make a loop here and just continually check if we can get a status. This is sub-optimal to say the least and if nothing else, instead of `Exception` it should use a better class.